addons/purchase/views/purchase_views.xml

```diff
@@ -181,15 +181,30 @@
                     </group>
                     <notebook>
                         <page string="Products">
-                            <field name="order_line" attrs="{'readonly': [('state', 'in', ('done', 'cancel'))]}">
-                                 <tree string="Purchase Order Lines" editable="bottom">
+                            <field name="order_line"
+                                widget="section_and_note_one2many"
+                                attrs="{'readonly': [('state', 'in', ('done', 'cancel'))]}">
+                                <tree string="Purchase Order Lines" editable="bottom">
+                                    <control>
+                                        <create name="add_product_control" string="Add a product"/>
+                                        <create name="add_section_control" string="Add a section" context="{'default_display_type': 'line_section'}"/>
+                                        <create name="add_note_control" string="Add a note" context="{'default_display_type': 'line_note'}"/>
+                                    </control>
+                                    <field name="display_type" invisible="1"/>
                                     <field name="currency_id" invisible="1"/>
                                     <field name="state" invisible="1"/>
                                     <field name="product_type" invisible="1"/>
                                     <field name="product_uom_category_id" invisible="1"/>
                                     <field name="sequence" widget="handle"/>
-                                    <field name="product_id" attrs="{'readonly': [('state', 'in', ('purchase', 'to approve','done', 'cancel'))]}" context="{'partner_id':parent.partner_id, 'quantity':product_qty,'uom':product_uom, 'company_id': parent.company_id}" force_save="1"/>
-                                    <field name="name" optional="show"/>
+                                    <field
+                                        name="product_id"
+                                        attrs="{
+                                            'readonly': [('state', 'in', ('purchase', 'to approve','done', 'cancel'))],
+                                            'required': [('display_type', '=', False)],
+                                        }"
+                                        context="{'partner_id':parent.partner_id, 'quantity':product_qty,'uom':product_uom, 'company_id': parent.company_id}"
+                                        force_save="1"/>
+                                    <field name="name" widget="section_and_note_text" optional="show"/>
```

Comment:
i'm waiting for confirmation from @fgi-odoo to know if we keep the optional="show" for the name field since it makes sections and notes lines look completely stupid if the user hides the field

Reply:
#34801 does the same in accounting FYI.

Reply:
confirmed, remove the optional=show thingy